tools/docgen/gen/doc_generator.go

```diff
@@ -0,0 +1,321 @@
+			err = gen.genCompositeDeclaration(astDecl)
+		case *ast.InterfaceDeclaration:
+			err = gen.genInterfaceDeclaration(astDecl)
+		default:
+			// do nothing
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (gen *DocGenerator) genCompositeDeclaration(declaration *ast.CompositeDeclaration) error {
+	if declaration.DeclarationKind() == common.DeclarationKindEvent {
+		return nil
+	}
+
+	declName := declaration.DeclarationIdentifier().String()
+	return gen.genCompositeDecl(declName, declaration.Members, declaration)
+}
+
+func (gen *DocGenerator) genInterfaceDeclaration(declaration *ast.InterfaceDeclaration) error {
+	declName := declaration.DeclarationIdentifier().String()
+	return gen.genCompositeDecl(declName, declaration.Members, declaration)
+}
+
+func (gen *DocGenerator) genCompositeDecl(name string, members *ast.Members, decl ast.Declaration) error {
+	gen.typeNames = append(gen.typeNames, name)
+
+	defer func() {
+		gen.typeNames = gen.typeNames[:len(gen.typeNames)-1]
+	}()
+
+	fileName := fmt.Sprint(gen.currentFileName(), mdFileExt)
+	f, err := gen.fileWriter(fileName)
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	err = gen.compositePageGen.Execute(f, decl)
+	if err != nil {
+		return err
+	}
+
+	return gen.genDeclarations(members.Declarations())
+}
+
+func (gen *DocGenerator) fileWriter(fileName string) (io.WriteCloser, error) {
+	if gen.files == nil {
+		return os.Create(path.Join(gen.outputDir, fileName))
+	}
+
+	return NewInMemoryFileWriter(gen.files, fileName), nil
+}
+
+func (gen *DocGenerator) currentFileName() string {
+	return strings.Join(gen.typeNames, nameSeparator)
+}
+
+var functions = template.FuncMap{
+	"isFunction": func(declaration ast.Declaration) bool {
+		return declaration.DeclarationKind() == common.DeclarationKindFunction
+	},
+
+	"isComposite": func(declaration ast.Declaration) bool {
+		switch declaration.DeclarationKind() {
+		case common.DeclarationKindStructure,
+			common.DeclarationKindStructureInterface,
+			common.DeclarationKindResource,
+			common.DeclarationKindResourceInterface,
+			common.DeclarationKindContract,
+			common.DeclarationKindContractInterface:
+			return true
+		default:
+			return false
+		}
+	},
+
+	"hasConformance": func(declaration ast.Declaration) bool {
+		switch declaration.DeclarationKind() {
+		case common.DeclarationKindStructure,
+			common.DeclarationKindResource,
+			common.DeclarationKindContract,
+			common.DeclarationKindEnum:
+			return true
+		default:
+			return false
+		}
+	},
+
+	"isInterface": func(declaration ast.Declaration) bool {
+		switch declaration.DeclarationKind() {
+		case common.DeclarationKindStructureInterface,
+			common.DeclarationKindResourceInterface,
+			common.DeclarationKindContractInterface:
+			return true
+		default:
+			return false
+		}
+	},
+
+	"isEnum": func(declaration ast.Declaration) bool {
+		return declaration.DeclarationKind() == common.DeclarationKindEnum
+	},
```

Comment:
Nice! We could maybe add these template function right to `DeclarationKind`, e.g. `DeclarationKind.IsTypeDeclaration` is already a similar one

Reply:
I wasn't sure whether it would be useful for other components, so didn't want to pollute the existing APIs.
Maybe we can add if there are more requirements?